File: addons/guide-cpp/src/modifiers/guide_modifier_canvas_coordinates.cpp
#include "guide_modifier_canvas_coordinates.h"
#include <godot_cpp/classes/engine.hpp>
#include <godot_cpp/classes/scene_tree.hpp>
#include <godot_cpp/classes/viewport.hpp>
#include <godot_cpp/classes/window.hpp>
#include <godot_cpp/core/math.hpp>

using namespace godot;

void GUIDEModifierCanvasCoordinates::_bind_methods() {
    ClassDB::bind_method(D_METHOD("get_relative_input"), &GUIDEModifierCanvasCoordinates::get_relative_input);
    ClassDB::bind_method(D_METHOD("set_relative_input", "val"), &GUIDEModifierCanvasCoordinates::set_relative_input);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "relative_input"), "set_relative_input", "get_relative_input");
}

Vector3 GUIDEModifierCanvasCoordinates::_modify_input(Vector3 input, double delta, int value_type) const {
    if (!input.is_finite()) return Vector3(Math_INF, Math_INF, Math_INF);

    SceneTree *tree =  Object::cast_to<SceneTree>(Engine::get_singleton()->get_main_loop());
    if (!tree) return Vector3(Math_INF, Math_INF, Math_INF);
    Viewport *viewport = tree->get_root();
    Transform2D xform = viewport->get_canvas_transform().affine_inverse();
    Vector2 coords = xform.xform(Vector2(input.x, input.y));

    if (relative_input) {
        Vector2 origin = xform.xform(Vector2(0, 0));
        coords -= origin;
    }

    return Vector3(coords.x, coords.y, input.z);
}

bool GUIDEModifierCanvasCoordinates::is_same_as(const Ref<GUIDEModifier> &other) const {
    Ref<GUIDEModifierCanvasCoordinates> o = other;
    if (o.is_null()) return false;
    return o->get_relative_input() == relative_input;
}

File: addons/GuideCS/Modifiers/GUIDEModifierCanvasCoordinates.cs
using Godot;

namespace Guide;

[Tool]
[GlobalClass]
public partial class GUIDEModifierCanvasCoordinates : GUIDEModifier {
	private bool _relativeInput;
	[Export]
	public bool RelativeInput {
		get => _relativeInput;
		set {
			_relativeInput = value;
			EmitChanged();
		}
	}

	public override bool IsSameAs(GUIDEModifier other) {
		return other is GUIDEModifierCanvasCoordinates otherCanvas &&
			   RelativeInput == otherCanvas.RelativeInput;
	}

	public override Vector3 _ModifyInput(Vector3 input, float delta, GUIDEAction.GUIDEActionValueType valueType) {
		if (!input.IsFinite()) return Vector3.Inf;

		if (Engine.GetMainLoop() is not SceneTree tree) return Vector3.Inf;
		var viewport = tree.Root;
		var transform = viewport.CanvasTransform.AffineInverse();
		var coordinates = transform * new Vector2(input.X, input.Y);

		if (RelativeInput) {
			var origin = transform * Vector2.Zero;
			coordinates -= origin;
		}

		return new Vector3(coordinates.X, coordinates.Y, input.Z);
	}

	public override string _EditorName() => "Canvas coordinates";

	public override string _EditorDescription() => "Converts a position input in viewport coordinates (e.g. from the mouse position input)\ninto canvas coordinates (e.g. 2D world coordinates). Useful to get a 2D 'world' position.";
}
